### PR TITLE
build: Use dynamic engine for generate. 

### LINKED ISSUE #1079: Megatron Inference support for GRPO (Dynamic batching)
GRPO appears to [support megatron as the generation backend](https://github.com/NVIDIA-NeMo/RL/blob/main/nemo_rl/algorithms/grpo.py#L325), but then will break later in the script when trying to call `prepare_refit_info` on a NoneType.

I am curious about whether it's possible to use a TRT Model Optimizer quantized model with the Fake Quant layers for training, for QAT GRPO. I assumed this would be simpler with the Megatron backend since no refit would be required (where I imagine this would be more complex for vLLM which wouldn't support those layers).

So my main question is: what is the state of using Megatron as the generation backend? Should this be supported? (If so, I think this doubles as a bug report).

Additional questions are:
* Is there any loose guidance or comparison on generation speed for Megatron vs vLLM?
* Does my intention regarding QAT and being easier with Megatron for generation hold water or am I misunderstanding something?

### PR DESCRIPTION
**Pull request by shanmugamr1992**
# What does this PR do ?

Adds mcore dynamic engine generation support 

https://wandb.ai/shanmugamr/grpo-dev?nw=nwusershanmugamr (PLOT showing vllm and mcore having similar performance)

So final times :
The numbers are as follows

Generation alone :
Mcore - 20 seconds
VLLM - 8 seconds

End to end step times :
Mcore - 47 seconds
VLLM - 37 seconds

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

Closes #1079 

# Usage

```bash
srun --gpus-per-node 8 --time 04:00:00  --account coreai_dlalgo_llm --job-name coreai_dlalgo_llm:inference --partition interactive --container-image /lustre/fsw/portfolios/coreai/users/terryk/enroot-images/gitlab-master.nvidia.com/terryk/images/nemo-rl:big-version-bump-from-githubci-18672942438.squashfs  --container-mounts /PATH/TO//RL:/opt/nemo-rl --account coreai_dlalgo_llm -N 1 -J coreai_dlalgo_llm-multimodal:debug --gpus-per-node=8 --no-container-mount-home -p interactive --pty bash

export HF_HOME=/path/to/store/model/and/data
export TORCH_CUDA_ARCH_LIST='9.0 10.0'
export HF_TOKEN=<hf_token>

uv run examples/run_grpo_math.py --config examples/configs/grpo_math_1B_megatron.yaml policy.generation.backend=megatron 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

## Summary by CodeRabbit

* **New Features**
  * Introduced dynamic inference engine for improved generation performance with CUDA graph optimization support.
  * Added configuration for GRPO Llama 3.2 1B Instruct model with Megatron generation backend.

* **Bug Fixes**
  * Fixed potential error in policy generation initialization.

* **Tests**
  * Added new functional tests for GRPO Megatron generation workflow.
  * Enabled previously disabled generation tests.

### CODE REVIEW

File: nemo_rl/models/policy/megatron_policy_worker.py [hunk 1]
```diff
@@ -1782,40 +1781,120 @@ def generate(
...
         from megatron.core.inference.model_inference_wrappers.gpt.gpt_inference_wrapper import (
             GPTInferenceWrapper,
         )
-
-        inference_context = StaticInferenceContext.from_config(inference_wrapper_config)
-
+        from megatron.core.inference.sampling_params import SamplingParams
+
+        inference_max_batch_size = inference_wrapper_config.inference_max_requests
+        max_batch_size = inference_max_batch_size
```
[the remark this answers is not in the document]
Comment (terrykong) [reply nemo_rl/models/policy/megatron_policy_worker.py]: @shanmugamr1992 okat to remove?

File: nemo_rl/models/policy/megatron_policy_worker.py [hunk 2]
```diff
@@ -1782,40 +1781,120 @@ def generate(
...
+        )
         from megatron.core.inference.model_inference_wrappers.gpt.gpt_inference_wrapper import (
             GPTInferenceWrapper,
         )
-
-        inference_context = StaticInferenceContext.from_config(inference_wrapper_config)
-
+        from megatron.core.inference.sampling_params import SamplingParams
+
+        inference_max_batch_size = inference_wrapper_config.inference_max_requests
```
Comment (terrykong) [inline nemo_rl/models/policy/megatron_policy_worker.py]: is this option and the ones below it need to be configurable? or are these generally sane values? if they are sane, could you add a comment above each on why that particular magic number?

File: nemo_rl/models/policy/megatron_policy_worker.py [hunk 3]
```diff
@@ -1782,40 +1781,120 @@ def generate(
...
             tokenizer=self.megatron_tokenizer,
         )
-        inference_engine = StaticInferenceEngine(
-            text_generation_controller=text_generation_controller,
-            max_batch_size=self.cfg["generation_batch_size"],
+
+        text_generation_controller.inference_wrapped_model.model.config.cuda_graph_impl = "local"
+        dynamic_engine = DynamicInferenceEngine(
+            controller=text_generation_controller,
+            random_seed=1234,
```
Comment (terrykong) [inline nemo_rl/models/policy/megatron_policy_worker.py]: vllm seeding isn't perfect (restarts of the program will reuse the seed from before), but maybe we should correct this seed.

The vllm one is based on the node and rank, but it should probably also be based on the step, that way when we restart we can ensure the training is deterministic and not based on when the training failed

File: nemo_rl/models/policy/megatron_policy_worker.py [hunk 4]
```diff
@@ -1782,40 +1781,120 @@ def generate(
...
+        )
+        sampling_params = SamplingParams(
+            temperature=1.0,
+            top_k=0,
+            top_p=0.0,
+            return_segments=True,
+            return_log_probs=True,
+            num_tokens_to_generate=self.cfg["generation"]["max_new_tokens"]
             - data["input_ids"].size(1),
         )
```
[the remark this answers is not in the document]
Comment (terrykong) [reply nemo_rl/models/policy/megatron_policy_worker.py:1890]: @shanmugamr1992 seems like coderabbit is correct here. looks like the bottom one is the old state and can be cleaned up

File: nemo_rl/models/policy/megatron_policy_worker.py [hunk 5]
```diff
@@ -1782,40 +1781,120 @@ def generate(
...
-        # print(out)
+
+        input_ids = data["input_ids"]
+        prompt_tokens_tensor = input_ids
+        prompt_lengths_tensor = data["input_lengths"]
+        request_id = 0
+        for p, prompt_len in zip(
+            prompt_tokens_tensor, prompt_lengths_tensor, strict=True
+        ):
+            tokenized_prompt = p[:prompt_len]  # .cuda()
```
Comment (terrykong) [inline nemo_rl/models/policy/megatron_policy_worker.py]: dbg comment?

File: nemo_rl/models/policy/megatron_policy_worker.py [hunk 6]
```diff
@@ -469,6 +463,9 @@ def __init__(
         }
         self.dtype = dtype_map[self.cfg["precision"]]
 
+        # Initialize generation step counter for seed calculation
+        self.generation_step = 0
```
Comment (terrykong) [inline nemo_rl/models/policy/megatron_policy_worker.py]: ok to drop for now since this impl still has same issue as vllm one of not being dependent on training step. prob will complicate impl so ok to drop
